### ISSUE TITLE
Add type safty to create method

### ISSUE THREAD

**Issue by ahhshm** [typeorm/typeorm#8215]
### The Problem

Lets say I have an entity:
```
@Entity()
export class User {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({ type: "varchar", length: 50 })
  name: string;
}
```

here name is not nullable. but when I want to create a new instance of it using create method:
`const user = userRepo.create({})`
I pass an empty object and I don't get any error. but when I want to save that instance I get a non-null constraint error. It's nice to see a typescript error like "name property is required" or "{} is not assignable to ...".
### Relevant Database Driver(s)

| DB Type              | Relevant      |
| ---                  | ---           |
| `aurora-data-api`    | no            |
| `aurora-data-api-pg` | no            |
| `better-sqlite3`     | no            |
| `cockroachdb`        | no            |
| `cordova`            | no            |
| `expo`               | no            |
| `mongodb`            | no            |
| `mysql`              | no            |
| `nativescript`       | no            |
| `oracle`             | no            |
| `postgres`           | no            |
| `react-native`       | no            |
| `sap`                | no            |
| `sqlite`             | no            |
| `sqlite-abstract`    | no            |
| `sqljs`              | no            |
| `sqlserver`          | no            |

### Are you willing to resolve this issue by submitting a Pull Request?

 - ✖️ Yes, I have the time, and I know how to start.
 - ✖️ Yes, I have the time, but I don't know how to start. I would need guidance.
 - ✖️ No, I don’t have the time, but I can support (using donations) development.
 - ✅ No, I don’t have the time and I’m okay to wait for the community / maintainers to resolve this issue.

**Comment by nosyminotaur** [comment]
You are getting this behaviour because of this function - 
https://github.com/typeorm/typeorm/blob/a868078f8f4f63cc8d81a63e098d639ed12c6608/src/repository/Repository.ts#L86
Check out the function argument, which is `DeepPartial`, which makes all the fields in the entity optional when passing it the `create()`.
This is a similar behaviour to when instantiating the `User` class.
Check out this example - 
```
@Entity()
export class User {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({ type: "varchar", length: 50 })
  name: string;
}
```

Now if you instantiate a user like so -
`const user = new User()`, and try saving this directly, you'll get the same error.
This object and the one passed in `create()` aren't far off.
Thus, you're getting this error.

I'm not defending the piece of code, I'm just telling you why it is. If there's no alternative for now, it will be quicker to just create a new instance manually and then saving it.

**Comment by ahhshm** [comment]
@nosyminotaur  Thank you for your reply. I used Prisma before. in Prisma if I want to initialize a new record without specifying non-nullable columns I get a typescript error that says something like "this column is required". I expected the same functionality in TypeORM because it is "Type" orm and everything should be typed strictly. So if I try to save a user that doesn't have the name I should see a typescript error:
```
const user = new User() // OR const user = userRepo.create({})
userRepo.save(user) // I expected something like: {} is not assignable to {name: string}
```

I think it's what it is. TypeORM and Prisma are different in many ways but I'm still in Prisma world. I little bit disappointed but it's ok :)

**Comment by nosyminotaur** [comment]
Can I take a look at your tsconfig.json config?
Usually, if a field in a class isn't initialised in the constructor, it raises an error, like so -
<img width="808" alt="Screenshot 2021-09-28 at 10 50 40 AM" src="https://user-images.githubusercontent.com/41340243/135027946-533b014c-a482-430e-a46a-2cbe93296b82.png">

Which means that either you need to assign a value to it in the constructor or in the property declaration.
This might be helpful in narrowing down such errors. So for example, if you have the `User` class as above and you try to create an instance of it, Typescript will not let you.

**Comment by ahhshm** [comment]
I disabled the `strictPropertyInitialization` so I can define a property without initializing it and I have a good reason for that. I have a table with let's say 10 columns, so I have 10 properties in the entity. besides that, I have to add some properties for relations and etc. initializing 15 properties, either in constructor or property declaration is very nasty, so I disabled it.

I don't think it will help me. my goal was to create a new record with an object in a type-safe way. I want to create a method that accepts an object as a parameter and passes it to the `create` method. then I realized that in TypeORM there is no type declaration for an entity so the non-nullable fields are required and others are optional. As you said, even the `create` method accepts an object with all the entities fields but all of them are optional.

I have a solution. I can create an interface for each entity with the mentioned properties. for example:
```
interface UserFields {
    id?: number,
    name: string
}
```
then I can tell my method about the type of the object:
```
const createUser = (data: UserFields) = {
   ...
}
```
Prisma creates that interface automatically, I was looking for something similar in TypeORM. 

Again, it is what is it. TypeORM is not Prisma and I try to get used to it.
